caikit_tgis_backend/managed_tgis_subprocess.py

```diff
@@ -0,0 +1,371 @@
+    STOPPED = 0  # subprocess does not exist
+    BOOTING = 1  # attempting to load the model
+    READY = 2  # ready for inference requests
+
+
+# pylint: disable=too-many-instance-attributes
+class ManagedTGISSubprocess:
+    def __init__(
+        self,
+        model_path: str,
+        grpc_port: int = 50055,
+        http_port: int = 3000,
+        health_poll_timeout: Optional[float] = None,
+        bootup_poll_delay: Optional[float] = None,
+        load_timeout: Optional[float] = None,
+    ):
+        if bootup_poll_delay is None:
+            bootup_poll_delay = 1
+        if health_poll_timeout is None:
+            health_poll_timeout = 1
+        if load_timeout is None:
+            load_timeout = 30
+
+        # parameters of the TGIS subprocess
+        self._model_path = model_path
+        self._grpc_port = grpc_port
+        self._http_port = http_port
+        self._bootup_poll_delay = bootup_poll_delay
+        self._health_poll_timeout = health_poll_timeout
+        self._load_timeout = load_timeout
+
+        self._hostname = f"localhost:{self._grpc_port}"
+        self._mutex = Lock()
+        self._tgis_state: _TGISState = _TGISState.STOPPED
+        self._tgis_proc = None
+        self._tgis_client = None
+        self._wrapped_client = None
+        self._bootup_thread = None
+        self._bootup_exc = None
+
+    def __del__(self):
+        self.terminate()
+
+    def is_ready(self) -> bool:
+        """Return true if a request can be propagated to the subprocess"""
+        return self._tgis_state == _TGISState.READY and self._tgis_client
+
+    def get_client(self):
+        """Creates a gRPC client to the subprocess
+
+        Returns None if the subprocess is not ready and doesn't have an internal client
+        """
+        if self._wrapped_client:
+            return self._wrapped_client
+
+        if self._tgis_client:
+            # create a wrapped gRPC client stub that defers to self._tgis_client
+            self._wrapped_client = GenerationServiceStubWrapper(self)
+            return self._wrapped_client
+
+        log.warning("<MTS7717800W>", "get_client called with _tgis_client set to None")
+        return None
+
+    def launch(self):
+        """Launch the subprocess or restart it if it exists"""
+        with self._mutex:
+            self._launch()
+
+    def wait_until_ready(self, timeout=None) -> bool:
+        start_t = time.time()
+
+        # wait for the boot monitoring thread if we have one
+        if self._bootup_thread:
+            self._bootup_thread.join(timeout)
+
+        while True:
+            log.debug("wait_until_ready saw state [%s]", self._tgis_state)
+            if self._tgis_state == _TGISState.READY:
+                return True
+
+            if self._tgis_state == _TGISState.STOPPED:
+                # raise the bootup exception if we captured one
+                if self._bootup_exc:
+                    raise self._bootup_exc
+                return False
```

Comment:
What is the `bool` return type used for here?

Reply:
Hmm, yeah... its a bit awkward currently... I'll update this function to either return if ready or throw an exception if not instead of having some cases where it returns a boolean False.